### PR TITLE
[SPARK-27711][CORE] Unset InputFileBlockHolder at the end of tasks

### PR DESCRIPTION
**Pull request by jose-torres**
## What changes were proposed in this pull request?

Unset InputFileBlockHolder at the end of tasks to stop the file name from leaking over to other tasks in the same thread. This happens in particular in Pyspark because of its complex threading model.

## How was this patch tested?

new pyspark test

### CODE REVIEW

**Comment by zsxwing** [comment]
LGTM

**Comment by zsxwing** [comment]
ok to test

File: python/pyspark/sql/tests/test_functions.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 import sys
 
 from pyspark.sql import Row
+from pyspark.sql.types import *
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py]: Can we avoid wlidcard import

File: python/pyspark/sql/tests/test_functions.py [hunk 2]
```diff
@@ -278,6 +279,22 @@ def test_sort_with_nulls_order(self):
...
+
+        non_file_df = self.spark.range(0, 100, 1, 100).select(input_file_name().alias('file'))
+
+        results = non_file_df.collect()
+        self.assertTrue(len(results) == 100)
+
+        # [SC-12160]: if everything was properly reset after the last job, this should return
+        # empty string rather than the file read in the last job.
+        for result in results:
+            assert(result[0] == '')
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py]: we can stick to `self.assertEqual` for a better message.

File: python/pyspark/sql/tests/test_functions.py [hunk 3]
```diff
@@ -278,6 +279,22 @@ def test_sort_with_nulls_order(self):
...
+        rdd = self.sc.textFile('python/test_support/hello/hello.txt').map(lambda x: {'data': x})
+        df = self.spark.createDataFrame(rdd, StructType([StructField('data', StringType(), True)]))
+        df.select(input_file_name().alias('file')).collect()
+
+        non_file_df = self.spark.range(0, 100, 1, 100).select(input_file_name().alias('file'))
+
+        results = non_file_df.collect()
+        self.assertTrue(len(results) == 100)
+
+        # [SC-12160]: if everything was properly reset after the last job, this should return
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py]: What's SC-12160?

File: python/pyspark/sql/tests/test_functions.py [hunk 4]
```diff
@@ -278,6 +279,22 @@ def test_sort_with_nulls_order(self):
             df.select(df.name).orderBy(functions.desc_nulls_last('name')).collect(),
             [Row(name=u'Tom'), Row(name=u'Alice'), Row(name=None)])
 
+    def test_input_file_name_reset_for_rdd(self):
+        from pyspark.sql.functions import udf, input_file_name
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py]: I think we can just import on the top

File: python/pyspark/sql/tests/test_functions.py [hunk 5]
```diff
@@ -278,6 +279,22 @@ def test_sort_with_nulls_order(self):
...
+        rdd = self.sc.textFile('python/test_support/hello/hello.txt').map(lambda x: {'data': x})
+        df = self.spark.createDataFrame(rdd, StructType([StructField('data', StringType(), True)]))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py]: actually, you don't have to import types:

```
spark.createDataFrame(rdd, "data STRING")
```

File: python/pyspark/sql/tests/test_functions.py [hunk 6]
```diff
@@ -278,6 +279,22 @@ def test_sort_with_nulls_order(self):
...
+        df.select(input_file_name().alias('file')).collect()
+
+        non_file_df = self.spark.range(0, 100, 1, 100).select(input_file_name().alias('file'))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py]: seems like we don;t need an alias `file`.

and why don't we just use `spark.range(100)`?

**Review by HyukjinKwon — APPROVED** [review]
Some nits. looks good to me too

File: python/pyspark/sql/tests/test_functions.py (hunk 3 above)
Comment (dongjoon-hyun) [reply python/pyspark/sql/tests/test_functions.py]: +1 for @HyukjinKwon 's comment. Is this the internal issue tracker ID?
Could you update the PR, @jose-torres ?

**Comment by jose-torres** [comment]
Pushed comments. Sorry it took so long, I was on a trip.

**Review by jiangxb1987 — APPROVED** [review]

**Comment by jiangxb1987** [comment]
Thanks! Merged to master, please manually backport to 2.4!

**Pull request closed without merging**